File: 1330/d/d.cc
#include <cstdio>
#include <vector>

using namespace std;
typedef long long ll;

const int maxd = 32;

int main() {
  int t;
  scanf("%d", &t);
  while (t--) {
    int d, m;
    scanf("%d%d", &d, &m);
    vector<int> cnt(maxd);
    ll num = 2;
    int i = 1;
    while (num - 1 <= d) {
      cnt[i] = num >> 1;
      num <<= 1;
      i++;
    }
    cnt[i] = d - num / 2 + 1;
    ll ans = 1;
    for (int len = 1; len < maxd; ++len)
      ans = ans * (cnt[len] + 1) % m;
    ans = (ans - 1 + m) % m;
    printf("%lld\n", ans);
  }
}

File: 1330/d/d.py
t = int(input())
maxd = 32
for case_num in range(t):
    d, m = map(int, input().split(' '))
    cnt = [0] * maxd
    num = 2
    i = 1
    while num - 1 <= d:
        cnt[i] = num // 2
        num *= 2
        i += 1
    cnt[i] = d - num // 2 + 1
    ans = 1
    for l in range(1, maxd):
        ans = ans * (cnt[l] + 1) % m
    ans = (ans - 1 + m) % m
    print(ans)
